impl_flock.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// +build !windows,!linux
+
+package mutex
+
+import (
+	"os"
+	"syscall"
+
+	"github.com/juju/errors"
+)
+
+type impl struct {
+	fd int
+}
+
+func acquire(name string) (Releaser, error) {
+	fd, err := syscall.Open(flockName(name), syscall.O_CREAT|syscall.O_RDONLY, 0600)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	err = syscall.Flock(fd, syscall.LOCK_EX|syscall.LOCK_NB)
+	if err != nil {
+		syscall.Close(fd)
+
+		if err == syscall.EWOULDBLOCK {
+			return nil, errors.Trace(ErrLocked)
+		}
+		return nil, errors.Trace(err)
+	}
+
+	return &impl{fd: fd}, nil
+}
+
+// Release implements Releaser.
+func (i *impl) Release() error {
```

Comment:
I anticipate that the common scenario is a double-release (and I don't know how a real release would fail either); that doesn't seem like a good reason to take down the whole process.

I could probably be swayed into agreeing that an error *actually releasing* the lock might be nuke-worthy; but the common simple case is murphy's law (of interfaces): if it can be used wrong, it will be used wrong. I'm much more concerned that people will write code paths that rarely double-release than I am that release will *actually* fail...